app/queries/get_change_offer_options.rb

```diff
@@ -0,0 +1,84 @@
+class GetChangeOfferOptions
+  attr_accessor :user, :for_provider, :recruitment_cycle_year
+
+  # application_choice.offered_course (either current or the original)
+  def initialize(user:, current_course:)
+    @user = user
+    @for_provider = current_course.accredited_provider || current_course.provider
+    @recruitment_cycle_year = current_course.recruitment_cycle_year
+  end
+
+  def make_decisions_courses
+    Course
+      .joins(training_provider_make_decisions)
+      .joins(ratifying_provider_make_decisions)
+      .where(combine_user_and_provider_permissions)
+  end
+
+  def offerable_courses
+    make_decisions_courses
+      .where(
+        open_on_apply: true,
+        recruitment_cycle_year: recruitment_cycle_year,
+      )
+      .where(ratifying_provider_is_preserved)
+  end
+
+  def available_providers
+    Provider
+      .with(offerable_courses: offerable_courses)
+      .joins('INNER JOIN offerable_courses ON providers.id = offerable_courses.provider_id')
+      .distinct
```

Comment:
```suggestion
      .group('providers.id')
```

can we avoid using distinct if possible.

Reply:
`distinct` performs badly in some cases, but no sure it will here. I'll have a look at the SQL produced and do some testing.

Reply:
Switched it to `.group` just to be on the safe side. Will have another look once it's merged.